### ISSUE TITLE
Huntsman v2 Analog Support

### ISSUE THREAD

**Issue by WrathofthePast** [1kc/razer-macos#250]
Let me know if there is other Huntsman v2 support (did not see it in the version history). I am getting mine delivered today and I will let everyone know if it fully works in single-step actuation mode and if the adjustable actuation works (supposed to be delivered last Sunday). I don't think dual-step actuation programming to work, but it might let me use it if I configure that on Windows save the profile to the on-board storage. I will follow up on that and anything else that comes up.

**Comment by WrathofthePast** [comment]
I received the keyboard today and made sure to update to v0.3.6 as I saw it added Huntsman v2 mini support and enable "Use F1, F2, etc. keys as standard function keys on external keyboards".
Works:

- Spectrum all Devices syncs the keyboard with the Deathadder Elite and Core X Chroma.
- The nonstandard buttons: skip forwards, skip backwards, play/pause, mute, & volume wheel
- USB-A passthrough for both power and data
- All qwerty keys
- Number lock has its MacOS function (clear)
- Remapping all function keys including print screen, scroll lock, and pause (translates to F13, F14, & F15 respectively)
Note: Remapping keys may or may not work using the on-board storage and Synapse 3 on Windows and plugging back into MacOS

Does not work:

- The keyboards spectrum is faster than the mouse and eGPU, so the spectrums do not stay synced.
- Setting a Custom Color; however, it does not work for the Deatheradder Elite or Core X Chroma either and never has
- Unable to properly assign: F9's "On-the-fly Macro Record Key, F10's "gaming mode," & pause's "sleep mode"
- Home (will not let me remap in system preferences)

**Comment by WrathofthePast** [comment]
As of v0.3.7, setting a custom color for the Core X Chroma and Deathadder v2 now works - Deathadder Elite presumably would as well, but it died a couple weeks ago. 

>Setting a Custom Color; however, it does not work for the Deatheradder Elite or Core X Chroma either and never has

Updated Statement:

Setting a Custom Color

**Comment by dhobi** [comment]
Could you let me know if these two points are still not working for you with the latest version (0.3.8)?
- The keyboards spectrum is faster than the mouse and eGPU, so the spectrums do not stay synced.
- Setting a Custom Color; however, it does not work for the Deatheradder Elite or Core X Chroma either and never has

**Comment by WrathofthePast** [comment]
> The keyboards spectrum is faster than the mouse and eGPU, so the spectrums do not stay synced.

Yes, but note the only keyboard I have is a Huntsman v2 Analog which isn't supported yet.

> Setting a Custom Color; however, it does not work for the Deatheradder Elite or Core X Chroma either and never has

Setting a custom color on the Huntsman v2 Analog still does not work as it is not recognized in v0.3.8. Deathadder and Core X Chroma custom colors now work as I previously stated: 

> As of v0.3.7, setting a custom color for the Core X Chroma and Deathadder v2 now works - Deathadder Elite presumably would as well, but it died a couple weeks ago.
> 
> > Setting a Custom Color; however, it does not work for the Deatheradder Elite or Core X Chroma either and never has
> 
> Updated Statement:
> 
> Setting a Custom Color

**Comment by dhobi** [comment]
Aha, I see, so the Huntsman v2 Analog spectrum is just the default one which can indeed differ in speed.
Thanks for the clarification!

**Comment by WrathofthePast** [comment]
OpenRazer picked this up: openrazer/openrazer#1458

**Comment by WrathofthePast** [comment]
@dhobi How necessary is openrazer doing this first? I'll fund this on issuehunt if that'll speed this up. The issue went quiet on openrazer/openrazer#1458 and I did not receive a response when inquiring about it. It's driving me crazy having to turn off the backlight every night and it's out of sync with the spectrum on my other Razer devices. The analog features might be a challenge according to openrazer, so I understand if there's nothing that can be done till openrazer figures it out. Would it be better to fund the issue over there?

**Comment by WrathofthePast** [comment]
@1kc Openrazer added support just over a month ago. I've been working on adding support for this device since then. I created a [fork](https://github.com/WrathofthePast/razer-macos) to work on it & uploaded what I currently have.  I'll create a pull request once it's working which could be as soon as tomorrow since I haven't tested it yet. However, I couldn't figure out how to do it with the new "Refactored MacOS C drivers into a submodule, which now lives in librazermacos" that came with v0.4.4, so I did this on v0.4.2. I may update it to v0.4.3-test, if you think that's just as stable. I'll be sure to let you know if my testing is successful & if not, what I need help with. Thank you for this wonderful program!

**Comment by 1kc** [comment]
@WrathofthePast fantastic! The refactor just means that you have to make changes to the c files in librazermacos instead of the drivers directory, you shouldn't have to make a lot of code changes. You would have to make a pull request to that as well. Let me know if you run into problems.

**Comment by WrathofthePast** [comment]
@1kc I attempted adding support to v0.4.5. How do I compile it since changes are now made to two different directories? I have yarn & developer tools installed.

**Comment by 1kc** [comment]
@overcurrent might be able to help, it used to be just be yarn dist

**Comment by WrathofthePast** [comment]
@1kc It wouldn't let me upload the razer-macos folder unless I compressed it since it's over 100 files. Is there a way around that since it doesn't compare the code to the master as a .zip upload? I presume paying for a GitHub subscription would remove that issue, so I'm asking about alternatives.

On a side note, this is done if it works, so I'm very eager to test this.

**Comment by 1kc** [comment]
Can you send the dmg file to me, then we can give it to people who want to test. Thanks

**Comment by WrathofthePast** [comment]
Ohhh, I was being really stupid. I downloaded the source code for librazermacos & razer-macos instead of cloning. I'll send the dmg in a moment.

**Comment by WrathofthePast** [comment]
@1kc I got a bit busy, but it's finally done and everything works. The original compiling bug I ran into had to do with iohook. I eventually gave up trying to fix that & manual development since that error didn't happen when using ./release.sh.

The only thing I didn't do was update the Razer MacOS version number. I figured that happens when you merge the pull request. 

I tried uploading the .dmg to the pull request, but it says the file must be <25 MB when it's 159.2 MB. Not sure how to get around that. Do I need a GitHub subscription to do it? If so, I have no problem paying for one of those. Normally, I would try to figure that out tonight on my own, but it's very late, so I'm going to bed. I'll deal with this upload stuff tomorrow which is now today, technically.
